src/analysis/normed/ring/seminorm.lean

```diff
@@ -31,33 +34,69 @@ set_option old_structure_cmd true
 
 open_locale nnreal
 
-variables {R S : Type*} (x y : R) (r : ℝ)
+variables {F R S : Type*} (x y : R) (r : ℝ)
 
 /-- A seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero, takes nonnegative
   values, is subadditive and submultiplicative and such that `f (-x) = f x` for all `x ∈ R`. -/
-structure ring_seminorm (R : Type*) [non_unital_ring R]
+structure ring_seminorm (R : Type*) [non_unital_non_assoc_ring R]
   extends add_group_seminorm R :=
 (mul_le' : ∀ x y : R, to_fun (x * y) ≤ to_fun x * to_fun y)
 
 /-- A function `f : R → ℝ` is a norm on a (nonunital) ring if it is a seminorm and `f x = 0`
   implies `x = 0`. -/
-structure ring_norm (R : Type*) [non_unital_ring R] extends add_group_norm R, ring_seminorm R
+structure ring_norm (R : Type*) [non_unital_non_assoc_ring R]
+  extends ring_seminorm R, add_group_norm R
+
+/-- A multiplicative seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero and
+multiplication, takes nonnegative values, is subadditive and such that `f (-x) = f x` for all `x`.
+-/
+structure mul_ring_seminorm (R : Type*) [non_assoc_ring R]
+  extends add_group_seminorm R, monoid_with_zero_hom R ℝ
+
+/-- A multiplicative norm on a ring `R` is a ring seminorm such that `f x = 0` implies `x = 0`. -/
```

Comment:
/-- A multiplicative norm on a ring `R` is a multiplicative ring seminorm such that `f x = 0` implies `x = 0`. -/